Compiler crash when index type not found

<!-- BUGS: Please use this template. -->
<!-- QUESTIONS: This is not a general support forum! Ask Qs at http://stackoverflow.com/questions/tagged/typescript -->
<!-- SUGGESTIONS: See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md -->

**TypeScript Version:**  2.1.1

**Code**

Compile the following file, with the compiler flag `-d`

```
export interface Test {
   [index: TypeNotFound]: any;
}
```


**Expected behavior:**

The following errors get printed to console, (errors copied from compiling without `-d` flag).

```
test.ts(3,6): error TS1023: An index signature parameter type must be 'string' or 'number'.
test.ts(3,13): error TS2304: Cannot find name 'TypeNotFound'.
```

**Actual behavior:**
The compiler crashes with the following stack trace

```
Error: Debug Failure. False expression: This is unknown parent for parameter: 154
    at Object.assert (/usr/local/lib/node_modules/typescript/lib/tsc.js:1697:23)
    at Object.fail (/usr/local/lib/node_modules/typescript/lib/tsc.js:1702:19)
    at getParameterDeclarationTypeVisibilityDiagnosticMessage (/usr/local/lib/node_modules/typescript/lib/tsc.js:44696:34)
    at Object.getParameterDeclarationTypeVisibilityError [as getSymbolAccessibilityDiagnostic] (/usr/local/lib/node_modules/typescript/lib/tsc.js:44643:41)
    at handleSymbolAccessibilityError (/usr/local/lib/node_modules/typescript/lib/tsc.js:43565:40)
    at emitEntityName (/usr/local/lib/node_modules/typescript/lib/tsc.js:43699:17)
    at emitTypeReference (/usr/local/lib/node_modules/typescript/lib/tsc.js:43715:17)
    at emitType (/usr/local/lib/node_modules/typescript/lib/tsc.js:43660:28)
    at writeTypeOfDeclaration (/usr/local/lib/node_modules/typescript/lib/tsc.js:43590:17)
    at emitParameterDeclaration (/usr/local/lib/node_modules/typescript/lib/tsc.js:44640:17)
```

--- a/src/compiler/declarationEmitter.ts
+++ b/src/compiler/declarationEmitter.ts
@@ -1625,6 +1625,12 @@ namespace ts {
                             Diagnostics.Parameter_0_of_call_signature_from_exported_interface_has_or_is_using_name_1_from_private_module_2 :
                             Diagnostics.Parameter_0_of_call_signature_from_exported_interface_has_or_is_using_private_name_1;
 
+                    case SyntaxKind.IndexSignature:
+                        // Interfaces cannot have parameter types that cannot be named
+                        return symbolAccessibilityResult.errorModuleName ?
+                            Diagnostics.Parameter_0_of_index_signature_from_exported_interface_has_or_is_using_name_1_from_private_module_2 :
+                            Diagnostics.Parameter_0_of_index_signature_from_exported_interface_has_or_is_using_private_name_1;
+
                     case SyntaxKind.MethodDeclaration:
                     case SyntaxKind.MethodSignature:
                         if (hasModifier(node.parent, ModifierFlags.Static)) {

--- a/src/compiler/diagnosticMessages.json
+++ b/src/compiler/diagnosticMessages.json
@@ -2300,6 +2300,14 @@
         "category": "Message",
         "code": 4090
     },
+    "Parameter '{0}' of index signature from exported interface has or is using name '{1}' from private module '{2}'.": {
+        "category": "Error",
+        "code": 4091
+    },
+    "Parameter '{0}' of index signature from exported interface has or is using private name '{1}'.": {
+        "category": "Error",
+        "code": 4092
+    },
 
     "The current host does not support the '{0}' option.": {
         "category": "Error",

--- a/tests/baselines/reference/declarationEmitIndexTypeNotFound.errors.txt
+++ b/tests/baselines/reference/declarationEmitIndexTypeNotFound.errors.txt
@@ -1,0 +1,17 @@
+tests/cases/compiler/declarationEmitIndexTypeNotFound.ts(3,6): error TS1023: An index signature parameter type must be 'string' or 'number'.
+tests/cases/compiler/declarationEmitIndexTypeNotFound.ts(3,13): error TS2304: Cannot find name 'TypeNotFound'.
+tests/cases/compiler/declarationEmitIndexTypeNotFound.ts(3,13): error TS4092: Parameter 'index' of index signature from exported interface has or is using private name 'TypeNotFound'.
+
+
+==== tests/cases/compiler/declarationEmitIndexTypeNotFound.ts (3 errors) ====
+    
+    export interface Test {
+        [index: TypeNotFound]: any;
+         ~~~~~
+!!! error TS1023: An index signature parameter type must be 'string' or 'number'.
+                ~~~~~~~~~~~~
+!!! error TS2304: Cannot find name 'TypeNotFound'.
+                ~~~~~~~~~~~~
+!!! error TS4092: Parameter 'index' of index signature from exported interface has or is using private name 'TypeNotFound'.
+    }
+    

--- a/tests/baselines/reference/declarationEmitIndexTypeNotFound.js
+++ b/tests/baselines/reference/declarationEmitIndexTypeNotFound.js
@@ -1,0 +1,9 @@
+//// [declarationEmitIndexTypeNotFound.ts]
+
+export interface Test {
+    [index: TypeNotFound]: any;
+}
+
+
+//// [declarationEmitIndexTypeNotFound.js]
+"use strict";

--- a/tests/cases/compiler/declarationEmitIndexTypeNotFound.ts
+++ b/tests/cases/compiler/declarationEmitIndexTypeNotFound.ts
@@ -1,0 +1,5 @@
+// @declaration: true
+
+export interface Test {
+    [index: TypeNotFound]: any;
+}

